### PR TITLE
feat: kandel

### PR DESCRIPTION
**Pull request by peterMangrove**

### CODE REVIEW

File: src/kandel.ts [hunk 1]
```diff
@@ -87,72 +84,139 @@ export function handlePair(event: Pair): void {
...
+  } 
+  kandelPopulateRetract.save();
+}
 
-export function handlePopulateStart(event: PopulateStart): void {}
+export function getOfferIdsForKandel(kandel: KandelEntity): string[] {
+  let offerIds = new Array<string>()
+  for (let i = 0; i < kandel.offerIndexes.length; i++) {
+    let info = kandel.offerIndexes[i].toString().split('-');
+      let offerNumber = info[1];
```
Comment (Louis-Amas) [inline src/kandel.ts:107]: Indentation is bad here

Comment (Louis-Amas) [reply src/kandel.ts:107]: You can also use const

File: src/kandel.ts [hunk 2]
```diff
@@ -87,72 +84,139 @@ export function handlePair(event: Pair): void {
...
+      let ba = info[2];
+      if (ba == '0') {
```
Comment (Louis-Amas) [inline src/kandel.ts]: What is '0' and '1', you can probably add a comment

File: src/kandel.ts [hunk 3]
```diff
@@ -87,72 +84,139 @@ export function handlePair(event: Pair): void {
...
+}
+
+export function handleRetractStart(event: RetractStart): void {
+  const kandelPopulateRetract = new KandelPopulateRetract(event.transaction.hash.toHex() );
+  kandelPopulateRetract.transactionHash = event.transaction.hash;
+  kandelPopulateRetract.creationDate = event.block.timestamp;
+  kandelPopulateRetract.isRetract = true;
+  kandelPopulateRetract.startLogIndex = event.logIndex;
+  kandelPopulateRetract.kandel = event.address;
+  kandelPopulateRetract.offerGives = [];
```
Comment (Louis-Amas) [inline src/kandel.ts:152]: Here it should be new Array()

File: src/mangrove-order.ts [hunk 4]
```diff
@@ -80,7 +80,7 @@ export function handleSetExpiry(event: SetExpiry): void {
   );
   const limitOrder = LimitOrder.load(offerId);
   if (!limitOrder) {
-    log.error("Missing limit order for offerId {}", [offerId]);
+    log.info("Missing limit order for offerId {}", [offerId]);
```
Comment (Louis-Amas) [inline src/mangrove-order.ts:83]: Why info ? why does it change in this PR ?

File: schema.graphql [hunk 5]
```diff
@@ -98,21 +99,43 @@ type Kandel @entity {
   reserveId: Account
   router: Account
 
-  totalBase: BigInt!
   depositedBase: BigInt!
-
-  totalQuote: BigInt!
   depositedQuote: BigInt!
 
+  offerIndexes: [Bytes!]!
```
Comment (Louis-Amas) [inline schema.graphql:105]: Why Bytes ? it's seems it's string

File: src/kandel.ts [hunk 6]
```diff
@@ -87,72 +84,139 @@ export function handlePair(event: Pair): void {
...
+  const kandel = KandelEntity.load(event.address)!; // TODO: use load in block
+  const offerIds = getOfferIdsForKandel(kandel);
+  const kandelPopulateRetract = KandelPopulateRetract.load( event.transaction.hash.toHex() )!; // TODO: use load in block
+
+  for(let i = 0; i < offerIds.length; i++) {
+    let offer = Offer.load(offerIds[i])!; // TODO: use load in block
+    if( offer.latestTransactionHash == event.transaction.hash && offer.latestLogIndex.gt( kandelPopulateRetract.startLogIndex) ) {
+      const totalGave = offer.totalGave === null ? BigInt.fromI32(0) : offer.totalGave;
+      const totalGot = offer.totalGot === null ? BigInt.fromI32(0) : offer.totalGot;
+      kandelPopulateRetract.offerGives = kandelPopulateRetract.offerGives.concat([ Bytes.fromUTF8( `${ offerIds[i] }-${ offer.gives }-${ totalGave.toString() }-${ totalGot.toString() }`) ]);
```
Comment (Louis-Amas) [inline src/kandel.ts:97]: Keep using string when doing concat, I've experienced errors with Bytes.

File: schema.graphql (hunk 5 above)
Comment (peterMangrove) [reply schema.graphql]: fixed

File: src/kandel.ts (hunk 3 above)
Comment (peterMangrove) [reply src/kandel.ts]: fixed

File: src/kandel.ts (hunk 2 above)
Comment (peterMangrove) [reply src/kandel.ts]: fixed

File: src/kandel.ts (hunk 1 above)
Comment (peterMangrove) [reply src/kandel.ts]: fixed

**Pull request merged**